### PR TITLE
HPCC-18702 Add HTTP Content-Encoding gzip/deflate support to ESP

### PR DESCRIPTION
**Pull request by mayx**
- Add methods to check if the content should be compressed or decompressed
- Add methods to compress or decompress the content
- Add flag to request and response to enable or disable compression, and
  set the flag properly on server and client sides.
- Add configuration entry to specify the minimum response content length
  to qualify for compression

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mayx** [comment]
@afishbeck please review...

File: esp/bindings/http/platform/httptransport.cpp [hunk 1]
```diff
@@ -2541,3 +2633,176 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+         || strnicmp(content_type, HTTP_TYPE_SOAP, strlen(HTTP_TYPE_SOAP)) == 0
+         || strnicmp(content_type, HTTP_TYPE_SVG_XML, strlen(HTTP_TYPE_SVG_XML)) == 0
+         || strnicmp(content_type, HTTP_TYPE_JSON, strlen(HTTP_TYPE_JSON)) == 0
+         || strnicmp(content_type, HTTP_TYPE_JAVASCRIPT, strlen(HTTP_TYPE_JAVASCRIPT)) == 0
+         || strnicmp(content_type, HTTP_TYPE_JAVASCRIPT2, strlen(HTTP_TYPE_JAVASCRIPT2)) == 0))
+        return false;
+
+    return true;
+#else
+    return false;
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: Because this simplifies everything, I would consider moving this to the very top.
bool CHttpResponse::shouldCompress(int& compressType)
{
#ifndef _USE_ZLIB
  return false;
#else
....
#endif

File: esp/bindings/http/platform/httptransport.cpp [hunk 2]
```diff
@@ -2541,3 +2633,176 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+    if (!context)
+        return false;
+    CHttpRequest* request = dynamic_cast<CHttpRequest*>(context->queryRequest());
+    if (!request)
+        return false;
+    StringBuffer acceptEncoding;
+    request->getHeader(HTTP_HEADER_ACCEPT_ENCODING, acceptEncoding);
+    if (acceptEncoding.length() == 0)
+        return false;
+    char* aestr = acceptEncoding.toLowerCase().detach();
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: Is it necessary to detach and free?  even if j_strok_r writes nulls it may still clean up in StringBuffer ok?

File: esp/bindings/http/platform/httptransport.cpp [hunk 3]
```diff
@@ -2541,3 +2633,176 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+    char* nextptr = nullptr;
+    bool found = false;
+    for (char* oneword = j_strtok_r(aestr, ", ", &nextptr); oneword; oneword = j_strtok_r(nullptr, ", ", &nextptr))
+    {
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: A more readable (although slightly less efficient) way of tokenizing is sometimes to:
StringArray compressTypes;
compressTypes.appendList(acceptEncoding.toLowerCase(), ",");

File: esp/bindings/http/platform/httptransport.cpp [hunk 4]
```diff
@@ -2541,3 +2633,176 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+        }
+    }
+    free(aestr);
+    if (!found)
+        return false;
+
+    const char* content_type = m_content_type.get();
+    if (m_content_type.length() == 0 ||
+         !( strnicmp(content_type, "text", 4) == 0
+         || strnicmp(content_type, HTTP_TYPE_APPLICATION_XML, strlen(HTTP_TYPE_APPLICATION_XML)) == 0
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: prefer strieq()

Comment (afishbeck) [reply esp/bindings/http/platform/httptransport.cpp]: Should we be ignoring trailing info after ";"?  For example Content-type: text/xml; charset="utf-8"?

File: esp/bindings/http/platform/httptransport.cpp [hunk 5]
```diff
@@ -1050,6 +1080,45 @@ StringBuffer& CHttpMessage::getHeader(const char* headername, StringBuffer& head
...
+    if (!headername || !*headername)
+        return false;
+
+    ForEachItemIn(x, m_headers)
+    {
+        const char* header = m_headers.item(x);
+        if (header == NULL)
+            continue;
+        const char* colon = strchr(header, ':');
+        if (colon == NULL)
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: nullptr?

File: esp/bindings/http/platform/httptransport.cpp [hunk 6]
```diff
@@ -1050,6 +1080,45 @@ StringBuffer& CHttpMessage::getHeader(const char* headername, StringBuffer& head
...
 
+bool CHttpMessage::hasHeader(const char* headername)
+{
+    if (!headername || !*headername)
+        return false;
+
+    ForEachItemIn(x, m_headers)
+    {
+        const char* header = m_headers.item(x);
+        if (header == NULL)
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: nullptr?

File: esp/bindings/http/platform/httptransport.cpp [hunk 7]
```diff
@@ -1050,6 +1080,45 @@ StringBuffer& CHttpMessage::getHeader(const char* headername, StringBuffer& head
...
+            continue;
+        const char* colon = strchr(header, ':');
+        if (colon == NULL)
+            continue;
+        unsigned len = colon - header;
+        if ((strlen(headername) == len) && (strnicmp(headername, header, len) == 0))
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: can calculate strlen(headername) once at the top since it never changes.

File: esp/bindings/http/platform/httptransport.cpp [hunk 8]
```diff
@@ -1112,6 +1181,14 @@ bool CHttpMessage::isFormSubmission()
         hasContentType(HTTP_TYPE_FORM_ENCODED));
 }
 
+void CHttpMessage::enableCompression()
+{
+    const char* off = getenv("ESP_COMPRESSION_OFF");
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: Remind me where all getenv values can be set?  Seem to remember differences in ESP, roxie, and thor.

File: esp/bindings/http/platform/httptransport.cpp [hunk 9]
```diff
@@ -2541,3 +2633,176 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+    if (!request)
+        return false;
+    StringBuffer acceptEncoding;
+    request->getHeader(HTTP_HEADER_ACCEPT_ENCODING, acceptEncoding);
+    if (acceptEncoding.length() == 0)
+        return false;
+    char* aestr = acceptEncoding.toLowerCase().detach();
+    char* nextptr = nullptr;
+    bool found = false;
+    for (char* oneword = j_strtok_r(aestr, ", ", &nextptr); oneword; oneword = j_strtok_r(nullptr, ", ", &nextptr))
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: Move this to common code like "zlibType2Header"?

**Review by afishbeck — COMMENTED** [review]
@mayx some comments.

**Comment by afishbeck** [comment]
@mayx

File: esp/bindings/http/platform/httptransport.cpp (hunk 8 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: The value can be passed in from command line shell. This is put in as a last-resort runtime fix for just in case, I don't expect it to ever be used (or hope that it will never be used).

File: esp/bindings/http/platform/httptransport.cpp (hunk 1 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Change made per suggestion.

File: esp/bindings/http/platform/httptransport.cpp (hunk 3 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Change made per suggestion.

File: esp/bindings/http/platform/httptransport.cpp (hunk 4 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Can't use strieq as this is partial match. All types that begins with text will be matched. Similarly all types starting with "application/xml" will be matched, so don't need to look at trailing info...

File: esp/bindings/http/platform/httptransport.cpp (hunk 5 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Change made per suggestion.

File: esp/bindings/http/platform/httptransport.cpp (hunk 6 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Change made per suggestion.

File: esp/bindings/http/platform/httptransport.cpp (hunk 7 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Change made per suggestion.

File: esp/bindings/http/platform/httptransport.cpp (hunk 9 above)
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: Change made per suggestion.

**Comment by mayx** [comment]
@afishbeck changes made per suggestions, please review again...

**Comment by afishbeck** [comment]
@wangkx can you take a look?  You were looking at some similar compression functionality in the past.

**Comment by wangkx** [comment]
@mayx have you fixed the NULL in line 1091 of esp/bindings/http/platform/httptransport.cpp? For some reason, I cannot see the fix?

**Comment by mayx** [comment]
@wangkx the changes were somehow stashed when I switched branches, now they're uploaded...

**Review by wangkx — COMMENTED** [review]
@mayx looks good. Just 4 minor comments/questions.

File: esp/bindings/http/platform/httptransport.cpp [hunk 10]
```diff
@@ -1050,6 +1080,46 @@ StringBuffer& CHttpMessage::getHeader(const char* headername, StringBuffer& head
...
+
+void CHttpMessage::removeHeader(const char* headername)
+{
+    if (!headername || !*headername)
+        return;
+
+    ForEachItemInRev(x, m_headers)
+    {
+        const char* header = m_headers.item(x);
+        if (header == NULL)
```
[the remark this answers is not in the document]
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: changed.

File: esp/bindings/http/platform/httptransport.cpp [hunk 11]
```diff
@@ -1050,6 +1080,46 @@ StringBuffer& CHttpMessage::getHeader(const char* headername, StringBuffer& head
...
+            continue;
+        const char* colon = strchr(header, ':');
+        if (colon == NULL)
```
[the remark this answers is not in the document]
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: changed

File: esp/bindings/http/platform/httptransport.cpp [hunk 12]
```diff
@@ -1112,6 +1182,14 @@ bool CHttpMessage::isFormSubmission()
         hasContentType(HTTP_TYPE_FORM_ENCODED));
 }
 
+void CHttpMessage::enableCompression()
+{
+    const char* off = getenv("ESP_COMPRESSION_OFF");
+    if (off && streq(off, "1"))
```
[the remark this answers is not in the document]
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: strToBool() interprets multiple values as true, here we just want it to be one value "1"...

File: esp/bindings/http/platform/httptransport.cpp [hunk 13]
```diff
@@ -2541,3 +2634,182 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+    catch(...)
+    {
+        IERRLOG("Unknown exception decompressing response content.");
+        return false;
+    }
+    if (originalContent != nullptr)
+        originalContent->setown(m_content);
+    m_content.setown(decompressedContent);
+    m_content_length = m_content.length();
+    removeHeader(HTTP_HEADER_CONTENT_ENCODING);
```
[the remark this answers is not in the document]
Comment (mayx) [reply esp/bindings/http/platform/httptransport.cpp]: @wangkx it's a close call, but I think it's slightly better/safer to remove the header because (1) the whole message has been transformed which includes content & content_length and (2) to avoid upper layers to do anything extra with the encoding header

**Comment by mayx** [comment]
@wangkx changes made per comments please review again...

**Review by wangkx — APPROVED** [review]
@mayx looks good.

**Comment by mayx** [comment]
@richardkchapman ready to be merged...

**Comment by richardkchapman** [comment]
@afishbeck Can you sanity-check please

File: esp/bindings/http/platform/httptransport.cpp [hunk 14]
```diff
@@ -2541,3 +2634,182 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
     if (handleExceptions(xslp, me, serv, meth, errorXslt) && logHandleExceptions)
         PROGLOG("Exception(s) handled");
 }
+
+inline bool zlibTypeFromHeader(const char* acceptEncodingHeader, int& compressType)
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: Don't know how often it's used, but there is the possiblity of weights after ;
Accept-Encoding: br;q=1.0, gzip;q=0.8, *;q=0.1
May at least want to strip them off before comparisons.

File: esp/bindings/http/platform/httptransport.cpp (hunk 4 above)
Comment (afishbeck) [reply esp/bindings/http/platform/httptransport.cpp]: true

**Review by afishbeck — COMMENTED** [review]
@mayx one comment about the possibility of "gzip;q=0.8," weighting modifiers in the accept encoding.

Otherwise looks good.

**Comment by mayx** [comment]
@afishbeck Processing of the weight modifier has been added, please review again. (don't think the smoke test error is related to this change as the change only affects ESP server side code)

File: esp/bindings/http/platform/httptransport.cpp [hunk 15]
```diff
@@ -2561,3 +2654,192 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
...
+    }
+
+    if (gzipq > 0 && gzipq >= deflateq && gzipq >= zlibq) //Use gzip as much as possible due to some reported browser issues with deflate
+        compressType = static_cast<int>(ZlibCompressionType::GZIP);
+    else if (deflateq > 0 && deflateq >= zlibq)
+        compressType = static_cast<int>(ZlibCompressionType::DEFLATE);
+    else if (zlibq > 0)
+        compressType = static_cast<int>(ZlibCompressionType::ZLIB_DEFLATE);
+
+    return gzipq > 0 || deflateq > 0 || zlibq > 0;;
```
Comment (afishbeck) [inline esp/bindings/http/platform/httptransport.cpp]: @mayx extra semicolon.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@mayx looks good, except an extra semicolon.

**Comment by mayx** [comment]
@afishbeck extra semicolon removed, please review again...

**Review by afishbeck — APPROVED** [review]
@mayx looks good.

**Comment by mayx** [comment]
@richardkchapman ready to merge (to master)...
